File: src/AttentionGuardian.Core/ScheduleTrial.cs
namespace AttentionGuardian.Core;

public static class ScheduleTrial
{
    public static ScheduleTrialResult Insert(
        IEnumerable<ScheduledTodo> existingSchedule,
        ScheduledTodo proposedTodo)
    {
        ArgumentNullException.ThrowIfNull(existingSchedule);
        ArgumentNullException.ThrowIfNull(proposedTodo);

        var existing = existingSchedule.ToArray();
        ValidateExistingSchedule(existing, proposedTodo);

        var ordered = existing
            .OrderBy(todo => todo.TimeRange.Start)
            .ThenBy(todo => todo.TimeRange.End)
            .ThenBy(todo => todo.Id)
            .ToArray();

        var mandatoryTodos = ordered
            .Where(todo => todo.IsMandatory)
            .ToArray();

        var conflicts = proposedTodo.IsMandatory
            ? mandatoryTodos
                .Where(todo => todo.TimeRange.Overlaps(proposedTodo.TimeRange))
                .Select(todo => new ScheduleConflict(
                    ScheduleConflictKind.MandatoryEventCollision,
                    proposedTodo,
                    todo))
                .ToArray()
            : [];

        var originalProposedTodo = proposedTodo;
        if (!proposedTodo.IsMandatory)
        {
            proposedTodo = MovePastMandatoryTodos(proposedTodo, mandatoryTodos);
        }

        var result = new List<ScheduledTodo>(ordered.Length + 1);
        var affected = new List<ScheduledTodo>();

        foreach (var todo in ordered)
        {
            if (todo.TimeRange.End <= proposedTodo.TimeRange.Start)
            {
                result.Add(todo);
            }
            else
            {
                affected.Add(todo);
            }
        }

        result.Add(proposedTodo);
        var occupiedUntil = proposedTodo.TimeRange.End;
        var chainIsActive = true;
        var hasRollover = proposedTodo.ScheduleDate > originalProposedTodo.ScheduleDate;

        foreach (var todo in affected)
        {
            if (!chainIsActive)
            {
                result.Add(todo);
                continue;
            }

            if (todo.IsMandatory)
            {
                if (todo.TimeRange.Start < occupiedUntil)
                {
                    occupiedUntil = todo.TimeRange.End > occupiedUntil
                        ? todo.TimeRange.End
                        : occupiedUntil;
                    result.Add(todo);
                    continue;
                }

                chainIsActive = false;
                result.Add(todo);
                continue;
            }

            if (todo.TimeRange.Start >= occupiedUntil)
            {
                chainIsActive = false;
                result.Add(todo);
                continue;
            }

            var shifted = MovePastMandatoryTodos(todo.MoveTo(occupiedUntil), mandatoryTodos);
            hasRollover |= shifted.ScheduleDate > todo.ScheduleDate;
            result.Add(shifted);
            occupiedUntil = shifted.TimeRange.End;
        }

        var finalSchedule = result
            .OrderBy(todo => todo.TimeRange.Start)
            .ThenBy(todo => todo.TimeRange.End)
            .ThenBy(todo => todo.Id)
            .ToArray();

        return ScheduleTrialResult.Success(finalSchedule, hasRollover, conflicts);
    }

    private static ScheduledTodo MovePastMandatoryTodos(
        ScheduledTodo todo,
        IReadOnlyList<ScheduledTodo> mandatoryTodos)
    {
        var moved = todo;

        foreach (var mandatoryTodo in mandatoryTodos)
        {
            if (moved.TimeRange.Overlaps(mandatoryTodo.TimeRange))
            {
                moved = moved.MoveTo(mandatoryTodo.TimeRange.End);
            }
        }

        return moved;
    }

    private static void ValidateExistingSchedule(
        IReadOnlyList<ScheduledTodo> existing,
        ScheduledTodo proposedTodo)
    {
        if (existing.Any(todo => todo is null))
        {
            throw new ArgumentException(
                "The existing schedule cannot contain null values.",
                nameof(existing));
        }

        if (existing.Any(todo => todo.Id == proposedTodo.Id))
        {
            throw new ArgumentException(
                "The proposed todo identifier must not already exist in the schedule.",
                nameof(proposedTodo));
        }

        if (existing.Select(todo => todo.Id).Distinct().Count() != existing.Count)
        {
            throw new ArgumentException(
                "The existing schedule cannot contain duplicate identifiers.",
                nameof(existing));
        }
    }
}

File: apple/AttentionGuardianDomain/Sources/AttentionGuardianDomain/ScheduleTrial.swift
import Foundation

public enum ScheduleTrialError: Error, Equatable {
    case duplicateIdentifier
}

public struct ScheduleConflict: Equatable, Sendable {
    public let proposedTodo: ScheduledTodo
    public let mandatoryTodo: ScheduledTodo
}

public struct ScheduleTrialResult: Equatable, Sendable {
    public let scheduledTodos: [ScheduledTodo]
    public let hasRolloverToNextDay: Bool
    public let conflicts: [ScheduleConflict]
}

public enum ScheduleTrial {
    public static func insert(
        _ proposedTodo: ScheduledTodo,
        into existingSchedule: [ScheduledTodo]
    ) throws -> ScheduleTrialResult {
        try validateIdentifiers(existingSchedule, proposedTodo: proposedTodo)

        let ordered = existingSchedule.sorted(by: stableScheduleOrder)
        let mandatoryTodos = ordered.filter(\.isMandatory)
        let conflicts = proposedTodo.isMandatory
            ? mandatoryTodos
                .filter { overlaps($0, proposedTodo) }
                .map {
                    ScheduleConflict(
                        proposedTodo: proposedTodo,
                        mandatoryTodo: $0)
                }
            : []

        let originalProposedTodo = proposedTodo
        let placedProposedTodo = proposedTodo.isMandatory
            ? proposedTodo
            : try movePastMandatoryTodos(proposedTodo, mandatoryTodos: mandatoryTodos)

        var result: [ScheduledTodo] = []
        var affected: [ScheduledTodo] = []
        for todo in ordered {
            if todo.end <= placedProposedTodo.start {
                result.append(todo)
            } else {
                affected.append(todo)
            }
        }

        result.append(placedProposedTodo)
        var occupiedUntil = placedProposedTodo.end
        var chainIsActive = true
        var hasRollover =
            placedProposedTodo.localDateOrdinal > originalProposedTodo.localDateOrdinal

        for todo in affected {
            guard chainIsActive else {
                result.append(todo)
                continue
            }

            if todo.isMandatory {
                if todo.start < occupiedUntil {
                    occupiedUntil = max(todo.end, occupiedUntil)
                    result.append(todo)
                } else {
                    chainIsActive = false
                    result.append(todo)
                }
                continue
            }

            if todo.start >= occupiedUntil {
                chainIsActive = false
                result.append(todo)
                continue
            }

            let shifted = try movePastMandatoryTodos(
                todo.moved(to: occupiedUntil),
                mandatoryTodos: mandatoryTodos)
            hasRollover =
                hasRollover || shifted.localDateOrdinal > todo.localDateOrdinal
            result.append(shifted)
            occupiedUntil = shifted.end
        }

        return ScheduleTrialResult(
            scheduledTodos: result.sorted(by: stableScheduleOrder),
            hasRolloverToNextDay: hasRollover,
            conflicts: conflicts)
    }

    private static func movePastMandatoryTodos(
        _ todo: ScheduledTodo,
        mandatoryTodos: [ScheduledTodo]
    ) throws -> ScheduledTodo {
        var moved = todo
        for mandatoryTodo in mandatoryTodos where overlaps(moved, mandatoryTodo) {
            moved = try moved.moved(to: mandatoryTodo.end)
        }
        return moved
    }

    private static func overlaps(
        _ left: ScheduledTodo,
        _ right: ScheduledTodo
    ) -> Bool {
        left.start < right.end && right.start < left.end
    }

    private static func stableScheduleOrder(
        _ left: ScheduledTodo,
        _ right: ScheduledTodo
    ) -> Bool {
        if left.start != right.start {
            return left.start < right.start
        }
        if left.end != right.end {
            return left.end < right.end
        }
        return left.id.uuidString.lowercased() < right.id.uuidString.lowercased()
    }

    private static func validateIdentifiers(
        _ existingSchedule: [ScheduledTodo],
        proposedTodo: ScheduledTodo
    ) throws {
        let identifiers = existingSchedule.map(\.id)
        guard Set(identifiers).count == identifiers.count,
              !identifiers.contains(proposedTodo.id) else {
            throw ScheduleTrialError.duplicateIdentifier
        }
    }
}
